docs/compaction-merge-loops.md

```diff
@@ -0,0 +1,166 @@
+
+Observations:
+- A kvset can have a `V` entry and a `PT` entry with the same key.
+
+## Condensed Input Stream
+
+We use the notion of a "condensed input stream" to be able to describe merge
+loops at a higher level.
+
+To understand what this means and why it is useful, consider the
+following two kvsets:
+```
+    Kvset 1:
+        Key    Values
+        ----   -----
+        foo1   V20, V16
+        foo2   V14
+        foo    PT18
+
+    Kvset 2:
+        Key    Values
+        ----   -----
+        foo1   T13, V12
+        foo2   V11
+
+Figure 2: Example input kvsets
+```
+
+In figure 1, step 2 produces a stream entries where each entry is a
+single entry from a single kvset.  We call this the raw intput stream.
+It would appear as follows for these example kvsets:
+```
+    get_next_raw() -->   [ foo,  [ PT18 ]]
+    get_next_raw() -->   [ foo1, [ V20, V16 ]]
+    get_next_raw() -->   [ foo1, [ T13, V12 ]]
+    get_next_raw() -->   [ foo2, [ V14 ]]
+    get_next_raw() -->   [ foo2, [ V11 ]]
+
+Figure 3: Raw input stream
+```
+
+The condensed view the same data is:
+```
+    get_next_condensed()  --> [ foo,  [ PT18 ]]
+    get_next_condensed()  --> [ foo1, [ V20, V16, T13, V12, PT18 ]]
+    get_next_condensed()  --> [ foo2, [ V14, V11, PT18 ]]
+
+Figure 4: Condensed input stream
+```
+
+Observations on the condensed view:
+
+- There has been no elimination of values.  It is simply a different
+  view of the entries contained in the two input kvsets.
+- Prefix tombstone ("foo", PT18) is associated with all keys that
+  match "foo".
+
+In the condensed view each key appears exactly once and with all of
+its values and all prefix tombstones that match the key.  The logic of
+translating from the raw stream to the condensed stream is not
+described here.
+
+## Compaction Merge Logic
+
+Using the condensed view abstraction, the general structure of the
+merge loop is simple:
+```
+1: void compact(iterator *iter, u64 horizon, bool drop_tombs) {
+2:
+3:     while (entry = iter->get_next_condensed()) {
+4:         key = entry.key();
+5:         values = entry.values();
+6:         decide which values should be kept;
+7:         if (any values are kept) {
+8:             output(key with kept values);
+9:         }
+10:    }
+
+Figure 5: An accurate compaction merge loop strucutre
+```
+
+All that is left to complete the description of compaction merge logic
+is to define the rules for value elimination logic (i.e., what happens
+at line 6 in figure 5).
+
+Some additional inputs to the loop are:
+
+- horizon: a lower bound on the view seqno of all active queries and
+  transactions
+- drop_tomb: true if and only if the merge loop should drop tombstones
+  and prefix tombstones
+
+Input drop_tomb is set to true when the input kvsets include the last kvset in a leaf node,
+in which case we know there are no older so the tombstones can be dropped.
+pppp
+Value elimination rules:
+
+1. Keep all values (V, T, and PT) with seqno > horizon.
```

Comment:
If you don't wrap the relation operator in backticks, I find it looks better if you just write out less than. Also wrap V, T, and PT in backticks